aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -1096,6 +1096,119 @@ Tensor clone(const Tensor& src, c10::optional<c10::MemoryFormat> optional_memory
   return self;
 }
 
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ covariations ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor cov(const Tensor& self, 
+           const c10::optional<Tensor>& otherOpt,
+           c10::optional<bool> rowvarOpt, 
+           c10::optional<bool> biasOpt, 
+           c10::optional<int64_t> ddofOpt,
+           const c10::optional<Tensor>& fweightsOpt, 
+           const c10::optional<Tensor>& aweightsOpt) {
+  
+  const Tensor other = otherOpt.value_or(Tensor());
+  const Tensor fweights = fweightsOpt.value_or(Tensor());
+  const Tensor aweights = aweightsOpt.value_or(Tensor());
+  auto bias = biasOpt.value_or(false);
+  auto ddof = ddofOpt.value_or(0);
+  auto rowvar = rowvarOpt.value_or(true);
+
+  if(self.size(0) == 0)
+    return at::empty_like(self);  
+  TORCH_CHECK(self.dim() <= 2,
+              "Input1 has more than 2 dimensions." );
+    
+  Tensor X, y;
+  if(self.dim() == 1)
+    X = self.unsqueeze(0);
+  else
+    X = self;
+
+  if(!rowvar && X.size(0) != 1)
+    X = X.t();
+
+  if(other.defined()) {
+    TORCH_CHECK(other.dim() <= 2,
+                "Input2 has more than 2 dimensions." );
+ 
+    if(other.dim() == 1)
+      y = other.unsqueeze(0);
+    else
+      y = other;
+
+    if(!rowvar && y.size(0) != 1)
+      y = y.t();
+
+    X = at::cat({X, y}, 0);
+  }
+
+  if (ddofOpt == c10::nullopt)
+    ddof = (bias == false) ? 1 : 0;
+  
+  Tensor w = at::empty(0, self.options());  
+  if(fweights.defined()) {
+    TORCH_CHECK((fweights.scalar_type() == at::kLong || fweights.scalar_type() == at::kInt),
+                "fweights must be integer.");
+    TORCH_CHECK(fweights.dim() <= 1,
+                "cannot handle multidimensional fweights.");
+    TORCH_CHECK(fweights.size(0) == X.size(1),
+                "incompatible numbers of samples and weights.");
+
+    TORCH_CHECK(fweights.min().item<float>() >= 0, 
```

Comment:
performing this check will synchronize a CUDA device with the CPU, which is expensive. 

cc @ngimel -- are we allowing syncs for a certain class of error checking now?

Reply:
Yes, we are doing similar checks in multinomial.